abra_core/src/typechecker/typechecker.rs

```diff
@@ -10,18 +10,24 @@ use std::iter::FromIterator;
 #[derive(Debug, PartialEq)]
 pub(crate) struct ScopeBinding(/*token:*/ Token, /*type:*/ Type, /*is_mutable:*/ bool);
 
+pub(crate) enum ScopeKind {
+    Block,
+    Function(/*token: */ Token, /*name: */ String),
+}
+
 pub(crate) struct Scope {
+    pub(crate) kind: ScopeKind,
```

Comment:
This is a mechanism for keeping track later on whether we're in a function scope, and we're referencing an identifier with that function's name: aka `recursion`